### ISSUE TITLE
AKS: Azure network policy addon blocks source-controller ingress

### ISSUE THREAD

**Issue by mazen-bassiouny** [fluxcd/flux2#703]
Dears, 

I'm trying to bootstrap flux2 in a new azure  AKS without any network policies defined.
After all CRDs are installed and the GitHub account is create, the bootstrap finishes with (time exceeded). The four controller pods are up and running. 
- The cluster is synchronized with the last commit correctly as below:

**#kubectl get gitrepositories.source.toolkit.fluxcd.io -A**
NAMESPACE     NAME          URL                                           READY   STATUS                                                            AGE
flux-system   flux-system   https://github.com/name/tanyflux   True    Fetched revision: main/6554ea6324d70caf0f2dfa200e137fd9c2aecc8a   59m

- But the Kustomization has the below error : 
**#kubectl get kustomizations.kustomize.toolkit.fluxcd -A**
failed to download artifact from http://source-controller.flux-system.svc.cluster.local./gitrepository/flux-system/flux-system/6554ea6324d70caf0f2dfa200e137fd9c2aecc8a.tar.gz, error: Get "http://source-controller.flux-system.svc.cluster.local./gitrepository/flux-system/flux-system/6554ea6324d70caf0f2dfa200e137fd9c2aecc8a.tar.gz": dial tcp 10.0.165.86:80: i/o timeout

- Same error exists while checking the Kustomization pod as below:
**#kubectl logs kustomize-controller-7f5455cd78-wwxhk -n flux-system**  
"level":"error","ts":"2021-01-14T09:04:52.524Z","logger":"controller","msg":"Reconciler error","reconcilerGroup":"kustomize.toolkit.fluxcd.io","reconcilerKind":"Kustomization","controller":"kustomization","name":"flux-system","namespace":"flux-system","error":"failed to download artifact from http://source-controller.flux-system.svc.cluster.local./gitrepository/flux-system/flux-system/6554ea6324d70caf0f2dfa200e137fd9c2aecc8a.tar.gz, error: Get \"http://source-controller.flux-system.svc.cluster.local./gitrepository/flux-system/flux-system/6554ea6324d70caf0f2dfa200e137fd9c2aecc8a.tar.gz\": dial tcp 10.0.165.86:80: i/o timeout"}

Thanks for any helpful advice.

**Comment by stefanprodan** [comment]
Duplicate of: #690 and #697

**Comment by frank-bee** [comment]
@stefanprodan 

with the latest cli I still get these errors

```
{"level":"info","ts":"2021-01-15T15:31:25.803Z","logger":"controller-runtime.metrics","msg":"metrics server is starting to listen","addr":":8080"}
{"level":"info","ts":"2021-01-15T15:31:25.805Z","logger":"setup","msg":"starting manager"}
I0115 15:31:25.806502       7 leaderelection.go:243] attempting to acquire leader lease  flux-system/7593cc5d.fluxcd.io...
{"level":"info","ts":"2021-01-15T15:31:25.806Z","msg":"starting metrics server","path":"/metrics"}
I0115 15:31:25.844814       7 leaderelection.go:253] successfully acquired lease flux-system/7593cc5d.fluxcd.io
{"level":"info","ts":"2021-01-15T15:31:25.907Z","logger":"controller.kustomization","msg":"Starting EventSource","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","source":"kind source: /, Kind="}
{"level":"info","ts":"2021-01-15T15:31:25.907Z","logger":"controller.kustomization","msg":"Starting EventSource","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","source":"kind source: /, Kind="}
{"level":"info","ts":"2021-01-15T15:31:26.007Z","logger":"controller.kustomization","msg":"Starting EventSource","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","source":"kind source: /, Kind="}
{"level":"info","ts":"2021-01-15T15:31:26.208Z","logger":"controller.kustomization","msg":"Starting Controller","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization"}
{"level":"info","ts":"2021-01-15T15:31:26.208Z","logger":"controller.kustomization","msg":"Starting workers","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","worker count":4}
{"level":"info","ts":"2021-01-15T15:32:02.963Z","logger":"controller.kustomization","msg":"Source not found","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","name":"flux-system","namespace":"flux-system"}
{"level":"error","ts":"2021-01-15T15:34:49.261Z","logger":"controller.kustomization","msg":"unable to send event","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","name":"flux-system","namespace":"flux-system","error":"POST http://notification-controller/ giving up after 5 attempt(s): Post \"http://notification-controller/\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"}
{"level":"info","ts":"2021-01-15T15:34:49.271Z","logger":"controller.kustomization","msg":"Reconciliation finished in 1m10.020911167s, next run in 10m0s","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","name":"flux-system","namespace":"flux-system","revision":"master/c1de637b168d45b34eb10bf67ec516403be5b82a"}
{"level":"error","ts":"2021-01-15T15:34:49.271Z","logger":"controller.kustomization","msg":"Reconciler error","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","name":"flux-system","namespace":"flux-system","error":"failed to download artifact from http://source-controller.flux-system.svc.cluster.local./gitrepository/flux-system/flux-system/c1de637b168d45b34eb10bf67ec516403be5b82a.tar.gz, error: Get \"http://source-controller.flux-system.svc.cluster.local./gitrepository/flux-system/flux-system/c1de637b168d45b34eb10bf67ec516403be5b82a.tar.gz\": dial tcp 10.0.200.230:80: i/o timeout"}


```

**Comment by frank-bee** [comment]
lastest cli == 0.6.0

**Comment by stefanprodan** [comment]
@frank-bee have you run `flux install` with 0.6?

**Comment by frank-bee** [comment]
with 0.6.1 

```
kubectl -n flux-system get networkpolicies.networking.k8s.io -oyaml
apiVersion: v1
items:
- apiVersion: networking.k8s.io/v1
  kind: NetworkPolicy
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"networking.k8s.io/v1","kind":"NetworkPolicy","metadata":{"annotations":{},"labels":{"app.kubernetes.io/instance":"flux-system","app.kubernetes.io/version":"latest"},"name":"allow-scraping","namespace":"flux-system"},"spec":{"ingress":[{"from":[{"namespaceSelector":{}}],"ports":[{"port":8080,"protocol":"TCP"}]}],"podSelector":{},"policyTypes":["Ingress"]}}
    creationTimestamp: "2021-01-16T08:46:24Z"
    generation: 1
    labels:
      app.kubernetes.io/instance: flux-system
      app.kubernetes.io/version: latest
    managedFields:
    - apiVersion: networking.k8s.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            .: {}
            f:kubectl.kubernetes.io/last-applied-configuration: {}
          f:labels:
            .: {}
            f:app.kubernetes.io/instance: {}
            f:app.kubernetes.io/version: {}
        f:spec:
          f:ingress: {}
          f:policyTypes: {}
      manager: kubectl-client-side-apply
      operation: Update
      time: "2021-01-16T08:46:24Z"
    name: allow-scraping
    namespace: flux-system
    resourceVersion: "2512"
    selfLink: /apis/networking.k8s.io/v1/namespaces/flux-system/networkpolicies/allow-scraping
    uid: 0742a3d4-15e2-4882-93e5-04a58e45b29b
  spec:
    ingress:
    - from:
      - namespaceSelector: {}
      ports:
      - port: 8080
        protocol: TCP
    podSelector: {}
    policyTypes:
    - Ingress
- apiVersion: networking.k8s.io/v1
  kind: NetworkPolicy
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"networking.k8s.io/v1","kind":"NetworkPolicy","metadata":{"annotations":{},"labels":{"app.kubernetes.io/instance":"flux-system","app.kubernetes.io/version":"latest"},"name":"allow-webhooks","namespace":"flux-system"},"spec":{"ingress":[{"from":[{"namespaceSelector":{}}]}],"podSelector":{"matchLabels":{"app":"notification-controller"}},"policyTypes":["Ingress"]}}
    creationTimestamp: "2021-01-16T08:46:24Z"
    generation: 1
    labels:
      app.kubernetes.io/instance: flux-system
      app.kubernetes.io/version: latest
    managedFields:
    - apiVersion: networking.k8s.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            .: {}
            f:kubectl.kubernetes.io/last-applied-configuration: {}
          f:labels:
            .: {}
            f:app.kubernetes.io/instance: {}
            f:app.kubernetes.io/version: {}
        f:spec:
          f:ingress: {}
          f:podSelector:
            f:matchLabels:
              .: {}
              f:app: {}
          f:policyTypes: {}
      manager: kubectl-client-side-apply
      operation: Update
      time: "2021-01-16T08:46:24Z"
    name: allow-webhooks
    namespace: flux-system
    resourceVersion: "2522"
    selfLink: /apis/networking.k8s.io/v1/namespaces/flux-system/networkpolicies/allow-webhooks
    uid: 9941d0f7-8490-4ac9-89ed-a1ae4a3e7614
  spec:
    ingress:
    - from:
      - namespaceSelector: {}
    podSelector:
      matchLabels:
        app: notification-controller
    policyTypes:
    - Ingress
- apiVersion: networking.k8s.io/v1
  kind: NetworkPolicy
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"networking.k8s.io/v1","kind":"NetworkPolicy","metadata":{"annotations":{},"labels":{"app.kubernetes.io/instance":"flux-system","app.kubernetes.io/version":"latest"},"name":"deny-ingress","namespace":"flux-system"},"spec":{"egress":[{}],"ingress":[{"from":[{"podSelector":{}}]}],"podSelector":{},"policyTypes":["Ingress","Egress"]}}
    creationTimestamp: "2021-01-16T08:46:24Z"
    generation: 1
    labels:
      app.kubernetes.io/instance: flux-system
      app.kubernetes.io/version: latest
    managedFields:
    - apiVersion: networking.k8s.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            .: {}
            f:kubectl.kubernetes.io/last-applied-configuration: {}
          f:labels:
            .: {}
            f:app.kubernetes.io/instance: {}
            f:app.kubernetes.io/version: {}
        f:spec:
          f:egress: {}
          f:ingress: {}
          f:policyTypes: {}
      manager: kubectl-client-side-apply
      operation: Update
      time: "2021-01-16T08:46:24Z"
    name: deny-ingress
    namespace: flux-system
    resourceVersion: "2524"
    selfLink: /apis/networking.k8s.io/v1/namespaces/flux-system/networkpolicies/deny-ingress
    uid: de49fb91-0b0b-4f1f-a2bd-84d624fe9df3
  spec:
    egress:
    - {}
    ingress:
    - from:
      - podSelector: {}
    podSelector: {}
    policyTypes:
    - Ingress
    - Egress
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""


```

**Comment by stefanprodan** [comment]
@frank-bee can you please try this: https://github.com/fluxcd/flux2/issues/697#issuecomment-760109352

**Comment by frank-bee** [comment]
that's what I just tries while you where posting the answer. worked :-)

**Comment by stefanprodan** [comment]
Hmm this must be a bug in Azure policy implementation, why does it bock traffic between pods in same namespace by default? Have you created some network policy that does that?

**Comment by frank-bee** [comment]
I think it should not, let me check

**Comment by frank-bee** [comment]
➜  gitops git:(master) kubectl get networkpolicies --all-namespaces
NAMESPACE     NAME                       POD-SELECTOR                  AGE
flux-system   allow-all-ingress-egress   <none>                        14m
flux-system   allow-scraping             <none>                        27m
flux-system   allow-webhooks             app=notification-controller   27m
flux-system   deny-ingress               <none>                        27m
kube-system   aks-link                   app=aks-link                  34m
➜  gitops git:(master)

**Comment by frank-bee** [comment]
```
kubectl -n kube-system get networkpolicies.networking.k8s.io -oyaml
apiVersion: v1
items:
- apiVersion: networking.k8s.io/v1
  kind: NetworkPolicy
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"networking.k8s.io/v1","kind":"NetworkPolicy","metadata":{"annotations":{},"labels":{"addonmanager.kubernetes.io/mode":"Reconcile"},"name":"aks-link","namespace":"kube-system"},"spec":{"egress":[{}],"podSelector":{"matchLabels":{"app":"aks-link"}},"policyTypes":["Egress"]}}
    creationTimestamp: "2021-01-16T08:40:05Z"
    generation: 1
    labels:
      addonmanager.kubernetes.io/mode: Reconcile
    managedFields:
    - apiVersion: networking.k8s.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            .: {}
            f:kubectl.kubernetes.io/last-applied-configuration: {}
          f:labels:
            .: {}
            f:addonmanager.kubernetes.io/mode: {}
        f:spec:
          f:egress: {}
          f:podSelector:
            f:matchLabels:
              .: {}
              f:app: {}
          f:policyTypes: {}
      manager: kubectl-client-side-apply
      operation: Update
      time: "2021-01-16T08:40:05Z"
    name: aks-link
    namespace: kube-system
    resourceVersion: "687"
    selfLink: /apis/networking.k8s.io/v1/namespaces/kube-system/networkpolicies/aks-link
    uid: 2bdca08a-8a71-4f4d-990f-0c6c2bde903f
  spec:
    egress:
    - {}
    podSelector:
      matchLabels:
        app: aks-link
    policyTypes:
    - Egress
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""


```

**Comment by frank-bee** [comment]
MS also writes here, that per default no rules are applied:
https://docs.microsoft.com/de-de/azure/aks/use-network-policies

**Comment by stefanprodan** [comment]
Ok so this must be a bug in Azure.

**Comment by stefanprodan** [comment]
@frank-bee can you please uninstall flux and install it with  --network-policy=false, just to confirm that Azure blocks traffic by default, this would help me determine if it’s something in our policy or it’s Azure. Thanks

**Comment by frank-bee** [comment]
I tested if azure blocks any network traffic  as described here
https://docs.microsoft.com/de-de/azure/aks/use-network-policies

I was possible to wget something from in the "frontend" accessing the "backend" in the  same namspace.
I think there is something wrong or incompatiple with azure  with the flux network policies.

**Comment by stefanprodan** [comment]
> I think there is something wrong or incompatiple with azure with the flux network policies.

Can you please install flux without its network policies to confirm this?

**Comment by frank-bee** [comment]
My daughter just woke up , I have to leaf for now. I can see if I find time next week,ok?

**Comment by frank-bee** [comment]
I re-installed flux on our other azure cluster with --network-policy=false
that worked

**Comment by frank-bee** [comment]
@stefanprodan  I tested re-installation now again on the recent azure AKS cluster (where I'm sure the azure policies are configured). As well, with ` --network-policy=false` install worked without the workaround (" allow all traffic" policy)

**Comment by stefanprodan** [comment]
If it works only with `--network-policy=false` then it must be something wrong with our policy https://github.com/fluxcd/flux2/blob/main/manifests/policies/deny-ingress.yaml :

```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: deny-ingress
spec:
  policyTypes:
    - Ingress
    - Egress
  ingress:
  - from:
    - podSelector: {}
  egress:
    - {}
  podSelector: {}
```

The role of this policy is block access from outside the namespace where Flux controllers are deployed.

**Comment by kingdonb** [comment]
I am currently working on getting a cluster with Azure CNI enabled so I can reproduce this myself. Will keep you posted.

Meanwhile, I was reviewing the NetworkPolicy docs at both Azure and Kubernetes, trying to understand the NetworkPolicy referenced above, and I found this specific policy that you described, mentioned in the Kubernetes upstream docs:

https://kubernetes.io/docs/concepts/services-networking/network-policies/#default-deny-all-ingress-and-all-egress-traffic

```yaml
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: default-deny-all
spec:
  podSelector: {}
  policyTypes:
  - Ingress
  - Egress
```

I will go ahead and try setting up Istio myself, so I have something to compare, but is there any reason the `deny-ingress` networkpolicy should be more complicated than this one? It might work better.

**Comment by stefanprodan** [comment]
@kingdonb network policies have nothing to do with Istio, so no need to install that.

**Comment by jonaskello** [comment]
Just to add another data-point, I had this exact same problem on an new AKS cluster today:

```
✔ install completed
► generating sync manifests
► applying sync manifests
◎ waiting for cluster sync
✗ failed to download artifact from http://source-controller.flux-system.svc.cluster.local./gitrepository/flux-system/flux-system/e9515d2f30bceb37cc611b9a296ff7973c9cdeda.tar.gz, error: Get "http://source-controller.flux-system.svc.cluster.local./gitrepository/flux-system/flux-system/e9515d2f30bceb37cc611b9a296ff7973c9cdeda.tar.gz": dial tcp 10.6.0.162:80: i/o timeout
```

I ran `flux uninstall` and then `flux boostrap .... --network-policy=false` and that fixed it.

**Comment by chanwit** [comment]
If I use the `deny-ingress` network policy meaning as @kingdonb pointed out,
> requests can get out from flux-system namespace but outside-in request to any flux-system service is blocked.

I suspect that the call from the kustomization controller to the source-controller is *NOT* simply pod-to-pod.
It's pod (kustomization-controller) -> service (source-controller) -> pod (source-controller), so it's an "outside-in" request and that's why it's blocked by the `deny-ingress` policy.

**Comment by stefanprodan** [comment]
@chanwit if that were true then Flux will not work with any other CNI. Can you confirm that on EKS/GKE+Calico it's the same problem?

**Comment by chanwit** [comment]
I did some testing on a local EKS-D + Calico cluster.
Flux worked nicely with network policy enabled.

To double check that the network policy was in action, I did curling from the default namespace to a service inside the flux-system ns, as suggested by @stefanprodan :

```
kubectl run curl-123 --image=radial/busyboxplus:curl -i --tty --rm
If you don't see a command prompt, try pressing enter.
[ root@curl-123:/ ]$ curl http://source-controller.flux-system.svc.cluster.local./gitrepository/flux-system/flux-system/9ad308e98eac2ca5c5cce6bdb8fbcb86c23db281.tar.gz
curl: (7) Failed to connect to source-controller.flux-system.svc.cluster.local. port 80: Connection timed out
[ root@curl-123:/ ]$
```

**Comment by stefanprodan** [comment]
Found this: https://github.com/Azure/AKS/issues/2031

It's nothing we can do about it, Azure AKS users should install flux with `--network-policy=false`.

**Comment by dtzar** [comment]
I've had success with multiple flux v2 bootstraps with AKS clusters, but I always use the `azure` network plugin.  This is generally recommended over the basic `kubenet`. 
See this article on [Azure CNI ](https://docs.microsoft.com/azure/aks/configure-azure-cni) and [AKS network policies](https://docs.microsoft.com/azure/aks/use-network-policies) for more info.

**Comment by kingdonb** [comment]
If `kubenet` can only support `NetworkPolicy` via Calico (or similar), this new info does not help clarify the issue for me, as according to the docs, I guess the Azure CNI network option can apparently use either Calico or Azure Network Policy.

The error report in Azure/AKS#2031 specifies that is related to Kubenet, but I believe in the configuration, either Azure CNI or Kubenet are mutually exclusive. You are either using Kubenet (eg. without network policies, or with network policies support provided via calico) or you are running Azure CNI (with network policy support provided natively, or provided via calico).

@dtzar can you clarify whether you are using Azure CNI with calico for networkpolicy, or the standalone CNI?

**Comment by dtzar** [comment]
@kingdonb Azure CNI only, no calico.  You are correct - you can only pick one of `kubenet` or `azure` for the AKS networking.

**Comment by chanwit** [comment]
Here's my setup which works for Flux.
```
az aks create --resource-group aks-chanwit  \
  --name aks-chanwit  --node-count 1  \
  --enable-addons monitoring  --generate-ssh-keys  \
  --network-plugin azure  \
  --network-policy calico
```

**Comment by kingdonb** [comment]
@dtzar are you sure you are not using calico for network policy? you mentioned azure for CNI only, this appears to be an issue when azure CNI is used with azure for networkpolicy. But if you are able to use flux with both networkPlugin and networkPolicy set to azure, there must be something different about your setup that permits this to work.

(It will help to clarify this for folks who are stuck with the broken behavior on azure/azure, if there is something different about your cluster... unless you really are using calico for policy as @chanwit shows above, which appears to work fine with Flux's network policy declarations!)

**Comment by dtzar** [comment]
Sorry after digging in a bit more I now realize I didn't enable `--network-policy`.  I only specified `--network-plugin azure` when I created my AKS cluster.

**Comment by hugo082** [comment]
Seems to have the same error with microk8s.

When I bootstrap my cluster I get the following error:
```
$ flux bootstrap ...
...
✗ failed to download artifact from http://source-controller.flux-system.svc.cluster.local./gitrepository/flux-system/flux-system/<HASH>.tar.gz, error: Get "...": dial tcp 10.152.183.240:80: connect: connection refused
```

I have tried with and without the `--network-policy=false` flag. I have also tried to apply the  [allow-all-ingress-egress](https://github.com/fluxcd/flux2/issues/697#issuecomment-760109352) network policy but none of those solutions work for me.  
Something strange, sometimes I get the `: dial tcp 10.152.183.240:80: i/o timeout` error instead of `: dial tcp 10.152.183.240:80: connect: connection refused` but I can't figure out when.

Before this issue, I had an issue that block requests from flux-system pods to kubernetes API service. This issue was solved with [this command](https://github.com/ubuntu/microk8s/issues/854#issuecomment-563886670). Maybe it has something to do with?

I'm using kubernetes `1.20` and flux `0.7.6`.

**Comment by chanwit** [comment]
MicroK8s didn't work for me too, but with the different error message.
```
► applying sync manifests
◎ waiting for cluster sync
✗ unable to clone 'ssh://git@github.com/chanwit/test-microk8s', error: dial tcp: lookup github.com on 127.0.0.53:53: read udp 127.0.0.1:53525->127.0.0.53:53: read: connection refused
```

Compared to MicroK8s, KIND's given me the best K8s on desktop experience so far.

**Comment by stefanprodan** [comment]
@hugo082 @chanwit this issue is about Azure. The dial tcp error in MicroK8s is not related to network policies, please read https://microk8s.io/docs/troubleshooting the "My pods can't reach the internet or each other" section.

**Comment by hugo082** [comment]
@stefanprodan I have read this section and my pods can reach the internet and each other. I have deployed a curl container and I was able to reach all services in my cluster.

Do you want me to open a new issue?

**Comment by chanwit** [comment]
@hugo082 opening a new one seems to be a good idea.

**Comment by stealthybox** [comment]
https://github.com/Azure/AKS/issues/2031 is marked as resolved with up-to-date AKS clusters.

We should test `flux bootstrap github` on the 5 AKS cni+netpol combinations to be sure before closing this.

1. --network-plugin=kubenet --network-policy=""
2. --network-plugin=kubenet --network-policy=calico
3. --network-plugin=azure --network-policy=""
4. --network-plugin=azure --network-policy=calico
5. --network-plugin=azure --network-policy=azure

azure netpol is not supported /w kubenet

**Comment by chanwit** [comment]
Please try no. 5 first as it's known to be broken.

**Comment by alanjcastonguay** [comment]
I have an AKS 1.18.14 cluster created 2021-03-18 with `--network-plugin=azure --network-policy=azure` on which I've deployed flux with using flux-cli version 0.9.1, without the `--network-policy=false` flag. I do not know whether this is sufficiently "up-to-date". I believe this is [case 5 @stealthybox described](#issuecomment-791517190).

Either I don't understand something about NetworkPolicy, or something's still broken. Or both.

The kustomize-controller is failing to connect to http://source-controller.flux-system.svc.cluster.local./. Observations:

* `curl` in either namespace `flux-system` or `default` also fails to connect to http://source-controller.flux-system.svc.cluster.local./.
* `curl` in either namespace `flux-system` or `default` can successfully connect to a placeholder Nginx container with `curl http://nginx.default.svc.cluster.local/`.
* `curl` in either namespace `flux-system` or `default` successfully makes connections to the source-controller pod on port 8080 (http-prom). This appears to be intentionally permitted by the NetworkPolicy `flux-system/allow-scraping`.

```
kubectl -n flux-system get endpoints source-controller
NAME                ENDPOINTS          AGE
source-controller   10.240.0.31:9090   3h51m
curl -s http://10.240.0.31:8080/metrics | head -n 1
# HELP controller_runtime_active_workers Number of currently used workers per controller
```

Creating the NetworkPolicy from https://github.com/fluxcd/flux2/issues/697#issuecomment-760109352 allows connections to source-controller from `curl` in either namespace `flux-system` or `default`. That seems slightly bad.

The only other NetworkPolicy resource present is for `tunnelfront`:

```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  labels:
    addonmanager.kubernetes.io/mode: Reconcile
  name: tunnelfront
  namespace: kube-system

spec:
  egress:
  - {}
  podSelector:
    matchLabels:
      component: tunnel
  policyTypes:
  - Egress
```

If instead I create this resource:

```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-flux-system-pods-to-use-source-controller-9090
  namespace: flux-system
spec:
  podSelector: {}
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          app.kubernetes.io/instance: flux-system
    ports:
    - port: 9090
```

* `curl` in namespace `flux-system` successfully to connect to http://source-controller.flux-system.svc.cluster.local./ and http://10.240.0.31:9090 (the source controller pod endpoint)
* `curl` in namespace `default` fails to connect to http://source-controller.flux-system.svc.cluster.local./. and http://10.240.0.31:9090 (the source controller pod endpoint)

Which appears to be _exactly what's originally intended_. Cool..?

Except those tests produce the same results after I delete the NetworkPolicy `allow-flux-system-pods-to-use-source-controller-9090`. Something seems wonky in Azure CNI reconciliation internals, at least from the perspective of my expectations. I'm not sure what else to test.

**Comment by eedorenko** [comment]
[The order of policy applying matters](https://www.danielstechblog.io/kubernetes-network-policies-on-azure-kubernetes-service-with-azure-npm/) for Azure Network policy. 
It appears that *allow-scraping* and *deny-ingress* policies fight with each other when Azure NP is enabled and as the result  *iptables* are configured wrong by *azure-npm* controller. But if you swap the order in *gotk-components.yaml*, so that *deny-ingress* is applied first, it works exactly as expected:
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  labels:
    app.kubernetes.io/instance: flux-system
    app.kubernetes.io/part-of: flux
    app.kubernetes.io/version: v0.12.0
  name: deny-ingress
  namespace: flux-system
spec:
  egress:
  - {}
  ingress:
  - from:
    - podSelector: {}
  podSelector: {}
  policyTypes:
  - Ingress
  - Egress
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  labels:
    app.kubernetes.io/instance: flux-system
    app.kubernetes.io/part-of: flux
    app.kubernetes.io/version: v0.12.0
  name: allow-webhooks
  namespace: flux-system
spec:
  ingress:
  - from:
    - namespaceSelector: {}
  podSelector:
    matchLabels:
      app: notification-controller
  policyTypes:
  - Ingress
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  labels:
    app.kubernetes.io/instance: flux-system
    app.kubernetes.io/part-of: flux
    app.kubernetes.io/version: v0.12.0
  name: allow-scraping
  namespace: flux-system
spec:
  ingress:
  - from:
    - namespaceSelector: {}
    ports:
    - port: 8080
      protocol: TCP
  podSelector: {}
  policyTypes:
  - Ingress
``` 

Tested on AKS 1.20.5 with --network-plugin=azure --network-policy=azure

**Comment by alanjcastonguay** [comment]
I don't expect apply-order to matter for most resources. Is this behavior of Azure Network Policy expected/desirable, or a bug?

**Comment by DrBushyTop** [comment]
There still seems to be some randomness in this functionality and I can't quite figure out what it is. I ran AKS `1.20.2` (with `--network-plugin=azure` and `--network-policy=azure`) for a week with flux  `0.13.1`. without any issues and yesterday this problem had surfaced. The age of my azure-npm pods is 25h and that's roughly when the problem started.

Tried updating to kubernetes `1.20.5` and flux `0.13.2`, removing the network policies and applying them again via flux install, but still can't seem to get this to run without setting `--network-policy=false` flag.

**Comment by nikkrg** [comment]
We are still seeing this issue. We are on Kubernetes 1.20.5 and we deploy flux using the terraform provider (version 0.1.6).

When `network plugin` and `network policy` both set to `azure` setting `network_policy` to `true` in the flux terraform provider causes the issue (kustomize controller couldn't connect to the source controller) whereas when not set to `true` everything works as expected. 

 Any ideas? Should we open a new issue?

**Comment by phillebaba** [comment]
Are you seeing these issues on a new cluster or are you reusing and older cluster?

**Comment by nikkrg** [comment]
@phillebaba we recreated the cluster a few days ago

**Comment by phillebaba** [comment]
Interesting, all the issues related to Network Policy and Azure CNI should be solved by now according to their GitHub issues. I saw this bug myself in AKS and it was solved by removing all faulty nodes, but this may not be the same issue if you are creating a new cluster. I would suggest that you contact Azure support as this is defiantly an AKS bug and not a Flux bug.
